extensions/data-plane/data-plane-api/src/main/java/org/eclipse/dataspaceconnector/dataplane/api/transfer/DataPlanePublicApiController.java

```diff
@@ -0,0 +1,113 @@
+
+/**
+ * Controller for synchronous data pull APIs of the data plane.
+ */
+@Consumes({MediaType.APPLICATION_JSON})
+@Produces({MediaType.APPLICATION_JSON})
+@Path("/transfer")
+public class DataPlanePublicApiController {
+
+    private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+    private final TokenValidationClient tokenValidationClient;
+    private final DataPlaneManager dataPlaneManager;
+    private final Monitor monitor;
+    private final TypeManager typeManager;
+
+    public DataPlanePublicApiController(TokenValidationClient tokenValidationClient, DataPlaneManager dataPlaneManager, Monitor monitor, TypeManager typeManager) {
+        this.tokenValidationClient = tokenValidationClient;
+        this.dataPlaneManager = dataPlaneManager;
+        this.monitor = monitor;
+        this.typeManager = typeManager;
+    }
+
+    @GET
+    @Path("validate/{token}")
+    public Response getData(@PathParam("token") String token, @Context UriInfo uriInfo, @Context Request request) {
+        // validate and decode input token
+        var tokenValidationResult = tokenValidationClient.validate(token);
+        if (tokenValidationResult.failed()) {
+            return notAuthorizedErrors(tokenValidationResult.getFailureMessages());
+        }
+
+        // create request and perform validation
+        var dataFlowRequest = createDataFlowRequest(tokenValidationResult.getContent(), uriInfo.getQueryParameters(), request.getMethod());
+        var validationResult = dataPlaneManager.validate(dataFlowRequest);
+        if (validationResult.failed()) {
+            return validationResult.getFailureMessages().isEmpty() ?
+                    validationError(String.format("Error processing  with EDR id: %s", dataFlowRequest.getId())) :
+                    validationErrors(validationResult.getFailureMessages());
+        }
+
+        // perform the data transfer
+        var stream = new ByteArrayOutputStream();
+        var sink = new OutputStreamDataSink(stream, executorService, monitor);
+        var transferResult = dataPlaneManager.transfer(sink, dataFlowRequest)
+                .exceptionally(throwable -> TransferResult.failure(ResponseStatus.FATAL_ERROR, "Unhandled exception: " + throwable.getMessage()))
+                .join();
+        if (transferResult.failed()) {
+            return internalErrors(transferResult.getFailureMessages());
+        }
+        return Response.ok().entity(Map.of("data", stream.toString())).build();
+    }
+
+    private DataFlowRequest createDataFlowRequest(ClaimToken claims, MultivaluedMap<String, String> queryParams, String method) {
+        var dataAddress = typeManager.readValue(claims.getClaims().get("dad"), DataAddress.class); // TODO: use constant here
+        return DataFlowRequest.Builder.newInstance()
+                .processId(UUID.randomUUID().toString()) // TODO: map the transfer process id into the token?
+                .destinationDataAddress(dataAddress)
```

Comment:
isn't it the sourceAddress instead of the destinationAddress that we have to provide?